Orm/Xtensive.Orm/Orm/Rse/ColumnCollection.cs

```diff
@@ -72,36 +106,29 @@ public ColumnCollection Alias(string alias)
     /// <summary>
     /// Returns an enumerator that iterates through the <see href="ColumnCollection"/>.
     /// </summary>
-    public List<Column>.Enumerator GetEnumerator() => columns.GetEnumerator();
+    public Enumerator GetEnumerator() => new Enumerator(this);
 
-    /// <inheritdoc/>
+    /// <summary>
+    /// Returns an enumerator that iterates through the <see href="ColumnCollection"/>.
+    /// </summary>
     IEnumerator<Column> IEnumerable<Column>.GetEnumerator() => GetEnumerator();
-    
+
     /// <inheritdoc/>
     IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
 
     // Constructors
 
-    /// <summary>
-    /// Initializes a new instance of this class.
-    /// </summary>
-    /// <param name="collection">Collection of items to add.</param>
-    public ColumnCollection(IEnumerable<Column> collection)
-      : this(collection.ToList())
-    {
-    }
-
     /// <summary>
     /// Initializes a new instance of this class.
     /// </summary>
     /// <param name="columns">Collection of items to add.</param>
-    public ColumnCollection(List<Column> columns)
+    public ColumnCollection(IEnumerable<Column> columns)
     {
-      this.columns = columns;
-      var count = columns.Count;
+      this.columns = columns as IReadOnlyList<Column> ?? columns.ToList();
```

Comment:
What about not having ```isinst``` check? RecordSetHeader can have ```IReadOnlyList``` instead of ```IEnumerable``` for ```columns``` parameter, right? I counted 3 places when we need to add ```.ToArray(count)```, enumeration is done anyway but we could make it more efficient by using ```.ToArray(count)``` instead of ```.ToList()``` here.

We still would have to have two ```ctors``` but having ready-to-go parameter is better than having computation. every time. Also as I see the only places where columns has to be given as IEnumerable are within ColumnCollection class.

Make sense?

Reply:
I agree.  Done